Firestore/core/src/firebase/firestore/local/query_data.h

```diff
@@ -0,0 +1,123 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_CORE_QUERY_DATA_H_
+#define FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_CORE_QUERY_DATA_H_
+
+#include <cstdint>
+#include <vector>
+
+#include "Firestore/core/src/firebase/firestore/core/query.h"
+#include "Firestore/core/src/firebase/firestore/model/snapshot_version.h"
+
+namespace firebase {
+namespace firestore {
+namespace local {
+
+/** An enumeration for the different purposes we have for queries. */
+enum class QueryPurpose {
+  /** A regular, normal query. */
+  kListen,
+
+  /**
+   * The query was used to refill a query after an existence filter mismatch.
+   */
+  kExistenceFilterMismatch,
+
+  /** The query was used to resolve a limbo document. */
+  kLimboResolution,
+};
+
+/**
+ * An immutable set of metadata that the store will need to keep track of for
+ * each query.
+ */
+class QueryData {
+ public:
+  /**
+   * Creates a new QueryData with the given values.
+   *
+   * @param query The query being listened to.
+   * @param target_id The target to which the query corresponds, assigned by the
+   *     LocalStore for user queries or the SyncEngine for limbo queries.
+   * @param purpose The purpose of the query.
+   * @param snapshot_version The latest snapshot version seen for this target.
+   * @param resume_token An opaque, server-assigned token that allows watching a
+   *     query to be resumed after disconnecting without retransmitting all the
+   *     data that matches the query. The resume token essentially identifies a
+   *     point in time from which the server should resume sending results.
+   */
+  QueryData(const core::Query& query,
+            int target_id,
+            QueryPurpose purpose,
+            const model::SnapshotVersion& snapshot_version,
+            const std::vector<uint8_t>& resume_token);
+
+  /**
+   * Convenience constructor for use when creating a QueryData for the first
+   * time.
+   */
+  // TODO(rsgowman): Define once WatchStream::EmptyResumeToken exists.
+  //QueryData(const core::Query& query, int target_id, QueryPurpose purpose);
+
+  const core::Query& query() const {
+    return query_;
+  }
+
+  int target_id() const {
+    return target_id_;
+  }
+
+  QueryPurpose purpose() const {
+    return purpose_;
+  }
+
+  const model::SnapshotVersion& snapshot_version() const {
+    return snapshot_version_;
+  }
+
+  const std::vector<uint8_t>& resume_token() const {
+    return resume_token_;
+  }
+
+  QueryData Copy(const model::SnapshotVersion& snapshot_version,
+                 const std::vector<uint8_t>& resume_token) const;
+
+ private:
+  const core::Query& query_;
```

Comment:
General nitpick: this class seems to be just a `struct` (no logic, a getter for each member variable) with all-`const` members. Would there be any difference in behavior if all its members were just `public`? Moreover, do its members even have to be `const`, when `Query` could just store it as a `const` member variable? I presume consistency with other platforms is a concern, though.

Reply:
QueryData are kept as values in maps making this non-reassignable means we're going to have to make the maps contain pointers to query data. Is that really what we want?

Reply:
Note that references are by design non-reassignable, so if this class were to be reassignable, I think they would have to become pointers.

Reply:
Converted to pointers; made non-pointers non-const. Switched back to class to allow me to hide that a bit.

IIUC, this should allow us to reassign it (thus allow it to work in containers) but keep the immutable interface.

Of course, we'll need to add a parameter-less ctor (which defines the ptrs to null) thus providing an invalid state. Hm. Maybe better to just use a map of pointers to query data instead. Or use docs to warn ourselves not to use the parameterless ctor except via maps.

thoughts?

Reply:
Both a null pointer and a default-constructed `QueryData` would serve the similar purpose of defining some sort of a blank state. I don't have an issue with having a default constructor (which is unavoidable for a value type), though I don't feel too strongly about it.

Reply:
Since there seem to be no further comments, I'll leave this as is, which should satisfy no one. :)

In it's current state, it's reassignable, but not default-constructable, so still can't be easily used in a map. (Proposal: Add the default ctor when we need it. Or don't, if we end up using a map of pointers.)

Reply:
> Proposal: Add the default ctor when we need it.

SGTM.